### ISSUE TITLE
Readonly data: Any equivalent in tablesaw to pandas' view vs copy?

### ISSUE THREAD

**Issue by mark27q1** [jtablesaw/tablesaw#296]
I've finally got my use case using tablesaw to an initial build and tried a few runs, and it is rather slow compared to a python version I wrote using pandas previously. I migrated to Java to get better concurrency in the hope of making it go faster.

Profiling it I see why -- it is spending 75% of its time in tech.tablesaw.table.Rows.copy(). This is because of some Table.where() calls that are designed to filter down the input data. 

Briefly, the background is my input data is price history for certain financial assets. Having established a price at a certain time (a simulated trade entry point) I then want to see if the market went up or down by a certain amount from that point. I am currently doing so using a filter like:

`useData.where(useData.numberColumn("High").isGreaterThanOrEqualTo(target).or(useData.numberColumn("Low").isLessThanOrEqualTo(stop)))`

(method names from memory so may have slight inaccuracies)

That gives me a filtered table, and then selecting the first row from this gives me the first instance that fulfilled the criteria. I'm only ever interested in the first row matching the criteria but can't think of a way to get tablesaw to stop once that row is found, so instead get the entire table and take the first row (or rather first value of each column) as needed. This kind of approach is applied repeatedly to further filter price data depending on what is found at each stage. The result is a lot of calls to Table.where().

As I say the code ends up spending a _lot_ of time copying rows, presumably from the original data to the return value of the where() method. Is there a way to get tablesaw to take a "view" approach as pandas would in this situation, that is not actually copy the data but simply copy references to the data, which would be faster? This obviously comes along with issues if one later tries to modify the filtered result represented by the view, because it would be impossible to do so without also modifying the original; pandas handles that by making attempts to alter a view an error, forcing the user to use an operation that would force a copy when that is what they want to do. For my particular use case, since the data is only read (in this particular case I don't even summarise it) I don't mind not being able to modify.

Does tablesaw have anything analogous to this view concept from pandas? Again, the goal is to avoid needlessly copying a lot of data.

**Comment by lwhite1** [comment]
> Does tablesaw have anything analogous to this view concept from pandas? 

sorta.  i'm not sure it will meet your needs, tho. It's only used internally for subtotals in summarization, so it may not currently support subsequent queries I will look and see what can be done. 
To use 'views':

        TableSlice slice = new TableSlice(sourceTable, aSelection);

This works with a reference to the original table and masks rows not indexed by aSelection.  My guess it is will require some work to make it suitable for your use-case, but it might be worth doing

> I'm only ever interested in the first row matching the criteria but can't think of a way to get tablesaw to stop once that row is found, so instead get the entire table and take the first row (or rather first value of each column) as needed.

I'm not sure how this leads to the next step in your process, so maybe this doesn't work for you, but it's easy to determine if any row meets the query criteria using Table::detect().  Here's an example (with a different table).
        
        Table t = Table.read().csv("../data/bush.csv");
        LocalDate dateTarget = LocalDate.of(2002, 1, 1);
        double ratingTarget = 75;
        Predicate<Row> test = row ->
                (row.getPackedDate("date").isAfter(dateTarget)
                && row.getInt("approval") > ratingTarget);
        boolean b = t.detect(test);

Sadly, there was a bug here so you need latest.  If you actually need the row value, you could just iterate the table using rows: 

       for (Row row : table) {
           if (row.getDouble("High") .... etc.
                  return t.rows(row.getRowNumber()); 
        }

Scanning is probably reasonably fast.  

Last but not least, if you read the same dataset over and over, it may be worthwhile to index it, then see if you get any hits.

**Comment by lwhite1** [comment]
Oh, i forgot this way. 
It may be possible to do what you want just using Selections. 
```
NumberColumn high = useData.numberColumn("High");
NumberColumn low = useData.numberColumn("Low");
Selection s = high.isGreaterThanOrEqualTo(target).or(low.isLessThanOrEqualTo(stop));

boolean anyMatches = ! s.isEmpty();

// get the first row that matches.
if (anyMatches) {
    Table firstMatchingRow = useData.where(s.get(0));    // Make sure selection is not empty first  
}

// do a subquery on the data that matches the first query;
Selection s2 = s.and(useData.numberColumn("x").isLessThan(1_000_000); 

// use s2 to get matches, or use in subqueries.

```

**Comment by mark27q1** [comment]
Absolutely brilliant -- both replies much appreciated. I will try the various approaches in the coming days and report back on results -- hopefully this will also be helpful to others. I also suspect some of these approaches will have a massively beneficial effect on memory use -- wasn't part of my original query but who doesn't like improved memory efficiency? 

And message received that the last reported is viewed as the most effective. I've got time on this project, and curious to see some stats on the different approaches. Will report back, and thanks again!

**Comment by mark27q1** [comment]
Absolutely brilliant -- both replies much appreciated. I will try the various approaches in the coming days and report back on results -- hopefully this will also be helpful to others. I also suspect some of these approaches will have a massively beneficial effect on memory use -- wasn't part of my original query but who doesn't like improved memory efficiency? 

And message received that the last reported is viewed as the most effective. I've got time on this project, and curious to see some stats on the different approaches. Will report back, and thanks again!

**Comment by lwhite1** [comment]
@mark27q1 You probably won't get too far with TableSlice as it exists, as it doesn't support queries. It's possible to implement them, but it would be a bit of work to do it right. There's also an open issue around what happens when you access their columns. 

Whether indexes or Selections is faster will depend on the details of your use case. indexes are faster once they exist, but can use a lot of memory (maybe involving garbage-collection-related performance hits). 

From my own testing, selections would be faster for some use-cases if they used real bitmaps rather than the compressed bitmaps they're built on, but they're pretty fast. They're also safer since memory usage doesn't grow as fast for complex queries.  It would be nice to offer real-bitmap-backed selections as an option for situations where very high performance is required and the memory usage is acceptable.

FWIW, I never considered the use of selections as a fast, low-memory way to handle table filtering in Tablesaw apps until you posted this. I will document it and probably use it regularly.

**Comment by mark27q1** [comment]
OK I have some results of my comparisons. 

First of all the code as quoted above doesn't quite work. Table.where() doesn't take an int, and selection.get() returns an int, so that code doesn't compile. However I found the rows() method of the Table class which _can_ take an int and that seems to do what I want. So:

`Table firstMatchingRow = useData.where(s.get(0));`

should be:

`Table firstMatchingRow = useData.rows(s.get(0));`

Now, to my results. First of all my test code is using as a base a dataset that is 1-minute price history (timestamp, open, high, low, close) of the GBPNZD currency pair from 2002 to 2013 October. That is each row is 1 minute's price action, showing where the market was at the beginning (open) and end (close) of that minute, and the highest and lowest prices that occurred in that minute. 

That constitutes 4,184,054 rows of data.

I then apply a hypothetical trading strategy to that data, simulating a trade management strategy that enters trades at each of a bunch of places in the price data starting at the beginning of 2006 and running up to the end of the data. The trading strategy is applied repeatedly to each of a set of pre-determined entry points; in the end the strategy is applied 975 times. The results of the hypothetical trades are output to a file at the end so I can compare and make sure my attempts to improve performance haven't broken the logic.

So that translates to a repeated filtering and decision-making function is applied to a 4,184,054-row data set 975 times in this test, each time sampling and making decisions according to a different part of the data.

My original method that used selections combined with Table.where() to filter the data repeatedly to make the decisions completed the analysis in **13 minutes and 18 seconds**.

Using selections as @lwhite1 suggested cut that to **4 minutes and 58 seconds**. 

Interestingly then introducing indexes (I need 3, on the datetime stamp column, on the high column and on the low column) and using the indexes to construct the selections, and behaving otherwise as in the selections version took **12 minutes and 5 seconds**. This implies that if I had used indexes and not applied the trick of using selections and then applying them to pull out only the row I was interested in it would have been even slower than the original version, which makes little sense to me. I am not completely convinced about those last results but I can't see anything obviously wrong with the code and each version produces identical results so nothing is horrendously broken.

The indexes took 7 seconds to construct, that 7 seconds is not counted in the above time.

So subject to me finding a mistake with my indexes-based version that explains its disappointing performance, it seems like the "use selections and only apply to a table when you absolutely have to" approach wins out by a long, long way -- even over "use indexes to build the selections and only apply to a table when you absolutely have to*. I do not understand why indexed columns are slower to construct selections on that un-indexed ones -- that doesn't appeal to my common sense -- but it is what I have found so far.

**Comment by mark27q1** [comment]
@lwhite1 I guess I am using Indexes correctly?

I construct an index on my data after loading it with:

`LongIndex tstampIndex = new LongIndex(priceData1M.dateTimeColumn("tstamp"));`

where my pricedata is in a Table variable priceData1M and that table has a DateTime column called tstamp which I constructed from the original data's Date and Time columns as they reside on disk.

Similar DoubleIndex construction lines construct the indexes on the High and Low columns.

Then when I use it a selection line like:

`Selection s = useData.dateTimeColumn("tstamp").isOnOrAfter(someDateTime).and(useData.numberColumn("High").isGreaterThanOrEqualTo(someHighValue).or(useData.numberColumn("Low").isLessThanOrEqualTo(someLowValue)));`

That line would become:

`Selection s = tstampIndex.atLeast(someDateTime).and(highIndex.atLeast(someHighValue).or(lowIndex.atMost(someLowValue)));`

Then in both cases I would find the first matching row with:

`Table candidate = useData.rows(s.get(0));`

Is this what you meant when you said use indexes?

**Comment by mark27q1** [comment]
Yes, actually, 4 minutes and 58 seconds to process 975 test runs of the strategy is 300ms per run, which really isn't too shabby at all. I didn't mean to imply otherwise. I know I didn't say much to this effect but I think that was a great improvement. What caught my attention more was why the use of **indexes (indices?)** actually _slowed down_ the query again to the point that it was almost back to where it started. That surprised me so much my first instinct was that I had messed something up, but looking at it again I don't think I did.

To answer your questions:

"Does the time include loading the data?" -- No, the data is loaded from disk once at the beginning, then repeatedly re-used. After the data is loaded a DateTime column is constructed from the loaded Date and Time columns, this takes a couple of seconds -- negligible time as far as I am concerned. The loaded columns are also named, since the file has no header -- that is really negligible time. Then, the index-using version makes the required indices. That takes 7 seconds. All that is done once. The times I quoted in my earlier posts are only the time taken by the series of tests. This steps through a list of LocalDateTimes, running a test of the strategy for each one, and then when it is finished skipping forward to the first that comes after the close of the just-finished trade, so none overlap. That is a simple for loop over a Java ArrayList so does not contribute substantially to the time. The vast majority of the time is spent in the actual strategy code, and most specifically in tablesaw methods relating to evaluating isOnOrAfter() or isGreaterThanOrEqualTo() or isLessThanOrEqualTo)0 (in the non-indexed cases -- I have timed but not profiled the indexed case).

"How fast was it using pandas?" -- Good question, I will have to check. My memory is that my full optimisation run, which involves repeating this test 19,000 times, would have taken in the region of 36 days to run in pandas if I had allowed it to, which was why I switched to Java so I could get parallelism to speak of, with the happy consequence that I discovered tablesaw! So 36 days divided by 19,000 (actually a bit more, but what's a few hundred between friends) is the answer to that. I will see if I can dig out more accurate test results for the pandas version shortly. The TLDR here is -- pandas was slow enough that I switched to Java looking for more speed (by making better use of my multi-core CPU)

"How fast does it need to be?" -- As fast as I can make it! I intend to construct myself a massive result dataset by running this test with each of a variation of 4 parameters. Each parameter is a numerical value that can vary between two range boundaries, eg between 1 and 10 in steps of 1, or between 100 and 1000 in steps of 50, and so on. I then wish to test the strategy for every combination of those 4 parameters and learn what relationships exist between the parameters. There are something a bit over 19,000 combinations. I'd ideally like that process not to take days -- but expect it will. My CPU has 8 cores so I'd like to be able to go 8 times as fast as pandas which only used 1. I'm just at the point now of trying to make it go as fast as possible. I'll be doing some different data science on the results once I get them!

"How much memory are you giving it?" -- my machine has 24GB and I have told the JVM it can use up to 20GB. With this initial test of a single instance of the strategy applied across the data (975 runs of the strategy) the maximum memory use I have seen (which occurs of course with the original version that did all that filtering and hence data copying) was 10GB. I'd ultimately like to test for other currency pairs than GBPNZD (which eventually means 19,000 * 28 tests, because I am interested in 28 currency pairs...) and so am hopeful that the selections-based enhancements you showed me will significantly reduce memory consumption, but I have not yet paid much attention to that metric.

"How many decimal places are used in the pricing data" -- the data is currency pricing from a retail broker, so being a non-JPY pair, it has 5 decimal places. Actually the brokers all used to use 4, and switched to 5 in 2005 or 2006. So the history shows X.XXXX0 before the change and Y.YYYYY after.

As for my patience -- you are helping me make this faster, so I find I can bring quite a lot of patience to bear! Having said that though I am travelling on business next week, and then am on the road a lot through the following month including some vacation time, so may go temporarily quiet here and there. Sure I am not alone in that... I am around for the rest of this week though.

**Comment by lwhite1** [comment]
Having thought on the problem more, I think the reason indexes are unhelpful is that you're working with inequalities on all three columns. It seems like the filters may have low selectivity, and so are of limited help.  

I have two suggestions. Both are based on doing a single scan of the table, rather than using selections, which does a scan for each query term. The code below is illustrative. Apologies if I broke it doing the cut and paste. 

The first approach should show improvement of 1/3 or so over the current best time, assuming I understand the problem well enough. It makes no assumptions about the data or the processing order.
 
First approach:

    int rowNumber = -1; 
    private static int getRowNumber1(Table t, LocalDateTime testDate, double testLow, double testHigh) {
        int rowNumber = -1;
        long testPackedDateTime = PackedLocalDateTime.pack(testDate);  // packing saves time
        for (Row row : t) {
            if (row.getPackedDateTime("date").isOnOrAfter(testPackedDateTime)
                    && (row.getDouble("value") > testLow
                    || row.getDouble("patient") > testHigh)) {
                return row.getRowNumber();
            }
        }
        return rowNumber;
    }
        /*
           get your row. Depending on what you use the row for, you could save a small amount of time 
           using the row directly, (by returning it from above, rather than the rowNumber,
           but just read the values and discard. 
        */

        if (rowNumber >= 0) 
             // do something with 
             t.rows(rowNumber);

The second approach assumes that (A) the data is sorted in ascending time order, and (B) the strategy tests are run in ascending time order.  If either condition does not hold, it would produce incorrect results. 

The basic idea is to reduce the number of rows that need to be scanned on each subsequent test by starting query n+1 on the row where query n found its first dateTime match.

Second approach: 
        // we use a field to maintain the startIndex position for the next run
        private int startIndex = 0;

// in main code

        // we are assuming that each test date is higher than the last, so we set them in this outer loop

        for (int i = 0; i < 975; i++) {
            testDateTime = testDateTime.plusDays(3);
            int rowNumber = getRowNumber(t, testDateTime, testValue, testPatient);
            if (rowNumber >= 0) {
                System.out.println(t.rows(rowNumber));
            }
        }

    private int getRowNumber(Table t, LocalDateTime testDate, double testLow, double testHigh) {
        int rowNumber = -1;
        boolean updatedStartIndex = false;
        long testPackedDateTime = PackedLocalDateTime.pack(testDate);  // packing saves time
        Row row = new Row(t);
        row.at(startIndex);
        while (row.hasNext()) {
            row.next();
            if (row.getPackedDateTime("date").isOnOrAfter(testPackedDateTime)) {
                if (!updatedStartIndex) {
                    startIndex = row.getRowNumber(); // update start index, we start searching here next time
                    updatedStartIndex = true;
                }
                if (row.getDouble("value") > testLow
                        || row.getDouble("patient") > testHigh) {
                    rowNumber = row.getRowNumber();
                    break;
                }
            }
        }
        return rowNumber;
    }

In the second approach you could again save some time by getting your values directly from the Row, rather than going back to the table.

**Comment by lwhite1** [comment]
So I understand, the 19,000 pandas runs would be 19,000 x the 975
iterations of the java test. Is that correct?  That’s actually faster than
your current best tablesaw numbers. (Looks around awkwardly)


"How fast was it using pandas?" -- Good question, I will have to check. My
> memory is that my full optimisation run, which involves repeating this test
> 19,000 times, would have taken in the region of 36 days to run in pandas
>

**Comment by mark27q1** [comment]
> So I understand, the 19,000 pandas runs would be 19,000 x the 975 iterations of the java test. Is that correct? That’s actually faster than your current best tablesaw numbers. (Looks around awkwardly)

Yes, _true_, but don't shuffle your feet too much -- the point of moving to Java was to be able to use a ThreadPool and really bring my CPU to bear on the problem, and I can still do that. And we haven't tried your latest suggested approach yet, and let's not forget Gawd alone knows what skulduggery pandas is doing under the hood... (well... _I_ don't know anyway!)

**Comment by mark27q1** [comment]
I will have a crack at your most recent suggestions either during today or, more likely, this evening (I'm in Japan). Will advise.

**Comment by lwhite1** [comment]
I think the second approach will be fast, and doesn’t actually require that
the strategies are applied in ascending date order.  There’s a way to do it
using a date index, assuming the table is date sorted.

YMMV, but I see this doing 1000 searches on 5 million rows:
> using rows with index. found 1000 in 78 ms

Which probably means I'm doing something wrong. ;)

**Comment by lwhite1** [comment]
For the record, this is the revised second approach, using an index to set the position to start scanning the db:
```
    // index on date to find the start position for searching    
   private LongIndex dateIndex;

    // in main code, we index
    dateIndex = new LongIndex(t.dateTimeColumn("date"));

    // run 1000 searches for different days
    for (int i = 0; i < 1000; i++) {
        testDateTime = testDateTime.plusDays(3);
        int rowNumber = getRowNumber(t, testDateTime, testValue, testPatient);
        // do something with the result.
    }

    private static int getRowNumber(Table t, LocalDateTime testDate, double testLow, double testHigh) {
        int rowNumber = -1;
        long testPackedDateTime = PackedLocalDateTime.pack(testDate);    // packing saves time
        Row row = new Row(t);
        row.at(dateIndex.get(testPackedDateTime).get(0));     // set the point to start scanning
        while (row.hasNext()) {
            row.next();
            if (row.getPackedDateTime("date").isOnOrAfter(testPackedDateTime)) {
                if (row.getDouble("lowValue") <= testLow
                        || row.getDouble("highValue") >= testHigh) {
                    rowNumber = row.getRowNumber();
                    break;
                }
            }
        }
        return rowNumber;
    }

```

**Comment by mark27q1** [comment]
So looking at approach number 2, I don't think that is going to work from outside tablesaw's package because Row has no public constructor any more -- you removed the "public" modifier from it in a commit a week or two back. I noticed because my early pre-testing version of my strategy code constructed a Row and that suddenly broke when I updated to latest. I responded by switching away from Row to using Table.xxxColumn().get() etc instead. Approach 1 should work because you can still get a Row out from an iterator, so for(Row row : t) should work. Will advise later if it does. 

For approach 2, do you think that's justification for re-instating the public modifier on Row's constructor? Or should we get our hands on a Row some other way?

**Comment by lwhite1** [comment]
I think you need to pull from master to get the public row constructor. I
put it back yesterday.

**Comment by benmccann** [comment]
I've released a new version with the public `Row` constructor

**Comment by lwhite1** [comment]
i saw that. Thank you!

**Comment by lwhite1** [comment]
@mark27q1 I'm going to close this, but feel free to open it if you're not completely satisfied with the answers.

**Comment by mark27q1** [comment]
I haven't been able to construct a test version using your last ideas yet, although I have spent some time thinking about how to. I expect to in the next few days and will report here. Based on what I find I'll leave it to your judgement whether to re-open or not -- I'm enjoying our collaboration on this but I'm aware this github project is about tablesaw not my trading projects so I'll understand if you feel we've gone deep enough.

**Comment by lwhite1** [comment]
No worries; work at your own pace and ask questions as you like. I get a
little anal sometimes about closing issues, but i've been enjoying this as
well. Much more fun than my day job ;)

**Comment by benmccann** [comment]
Ok, added to readme

**Comment by mark27q1** [comment]
OK I have some more results here. Same test doing the same set of runs across the same data set. I tried both the first and second methods above, except I did not use a static method since the index is an instance variable and the static method would not have been able to access it.

The first method drops the time to perform the test to 3 minutes and 45 seconds or so. In that method I do not use an index but pass the Row object in and out, searching the data sequentially. this requires that all the tests are performed in time order, but they are so that is fine.

The second method, using the index to find the starting point and then searching row by row for the trigger row, completes in 3 minutes and 25 seconds or so (approximation because different runs completed in slightly different times, plus or minus a few seconds).

So, using the index is marginally faster. I am wondering if a small additional benefit can be gained by using the index once per test to find the entry point, and then stepping from there until a test trade is complete. I will try that but if it makes a difference it will be a very small one.

So, barring any revelations from here, I think that is about as fast as I am going to get this, and the best approach for this seems to be the very last suggested approach above.

This was tested with the very latest tablesaw pulled from git about 20 minutes ago.

**Comment by mark27q1** [comment]
More results -- I found  both my implementations were still doing one call to Table.where for each trade in order to find the opening price for the trade. Eliminating that led to the following results:

1st approach (no index) -- 29 seconds.
2nd approach (used index) -- 39 seconds.

NOW we're cooking. That's quite a bit better than the original 13 minutes!

**Comment by lwhite1** [comment]
Cool. I did some more analysis and came to the conclusion that your code
was doing something I didn’t know about. TLDR you can scan the whole table
1000 times in about 70 secs, and performance is linear with average rows.
If you’re scanning a lot of rows you’re probably doing as well as you can.

**Comment by lwhite1** [comment]
@mark27q1 
BTW, I am curious to see how well this performs when you get the multi-threading work done. Please update when you've got some results if you have time.

**Comment by mark27q1** [comment]
Will do. I fired up the multi-threaded approach the other day and ran into an easily-solved problem which led to a suggestion I'll make on here in a minute in a new issue. In short I was trying to write CSV's generated from Tables out to standard output and using write().csv() closes its writer, which closes the OutputStream, which closes System.out after the first table has been written... So I was getting the results from the very first run to complete, and none of the others... I sorted that already.

Now I can see it chugs along very nicely, but I'm embarrassed to admit I am now worried about overheating my machine and causing a fire -- the machine is self-built and I may have undergunned the cooling a little, as I can see the temperature sensors getting close to the red-line when this has been running a few minutes. I'm thinking of migrating it to the cloud to do the full run. I've done plenty of hardcore parallel processing before but not that I have expected to run for this long. Another option is to use a smaller threadpool to leave a vCPU or two free, which may help keep the temperature down. Since I'm on the road and the machine is currently unattended, I'm being consciously paranoid here.

So no results yet, but I will say that using an approximation of 30 seconds per run and re-calculating the numbers for the number of runs I am expecting, a naive estimate for the expected run time is between 18 and 22 hours, which is quite a bit better than the 36 days I calculated for pandas...
